File: p3.py
from math import sqrt
# find largest prime factor of 600851475143

# Factor x:
# starting with 2, find a number that divides x.
# return that number with the others

def factors(x):
    for i in xrange(2, int(sqrt(x) + 1)):
        if x % i == 0:
            return [i] + factors(x / i)
    return [x]

print factors(600851475143)
print factors(10)

File: p3.hs
module Main where
main = do print ans

-- find the first prime factor of n, including n
factors n
    | head fs == n = n
    | otherwise    = factors (n `div` (head fs))
    where fs = dropWhile ((/= 0) . (mod n)) [2..floor (sqrt (fromIntegral n))] ++ [n]

ans = factors 600851475143
